### PR TITLE
refactor: Create class for column ids and column refs

### PR DESCRIPTION
**Pull request by TrevorBergeron**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery-dataframes/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> 🦕

### CODE REVIEW

File: bigframes/core/compile/compiled.py [hunk 1]
```diff
@@ -351,7 +356,15 @@ def _to_ibis_expr(
         return table
 
     def filter(self, predicate: ex.Expression) -> UnorderedIR:
-        if any(map(is_window, map(self._get_ibis_column, predicate.unbound_variables))):
+        if any(
+            map(
+                is_window,
+                map(
```
Comment (sycai) [inline bigframes/core/compile/compiled.py:362]: This if expression looks super long. Shall we just use a plain-old for loop?

for col in predicate.column_references:
  if (is_window(self._get_ibis_column(col.sql))):
      return ....

File: bigframes/core/compile/compiled.py [hunk 2]
```diff
@@ -1069,7 +1091,15 @@ def _to_ibis_expr(
         return table
 
     def filter(self, predicate: ex.Expression) -> OrderedIR:
-        if any(map(is_window, map(self._get_ibis_column, predicate.unbound_variables))):
+        if any(
```
Comment (sycai) [inline bigframes/core/compile/compiled.py:1094]: Probably can be replaced by a simple for loop too...

**Review by sycai — APPROVED** [review]

**Comment by sycai** [comment]
LGTM. I only have some style nits for your to consider.

File: bigframes/core/expression.py [hunk 3]
```diff
@@ -31,6 +32,10 @@ def const(
     return ScalarConstantExpression(value, dtype or dtypes.infer_literal_type(value))
 
 
+def deref_name(name: str) -> DerefOp:
```
[the remark this answers is not in the document]
Comment (TrevorBergeron) [reply bigframes/core/expression.py]: sounds good, done

File: bigframes/core/expression.py [hunk 4]
```diff
@@ -169,21 +188,57 @@ class UnboundVariableExpression(Expression):
...
+    def is_const(self) -> bool:
+        return False
+
+    def output_type(
+        self, input_types: dict[ids.ColumnId, bigframes.dtypes.Dtype]
+    ) -> dtypes.ExpressionType:
+        raise ValueError(f"Type of variable {self.id} has not been fixed.")
+
+    def bind_refs(
+        self, bindings: Mapping[ids.ColumnId, Expression], check_bind_all: bool = True
```
[the remark this answers is not in the document]
Comment (TrevorBergeron) [reply bigframes/core/expression.py]: yeah, that is clearer. converted all instance of this arg name

File: bigframes/core/identifiers.py [hunk 5]
```diff
@@ -11,16 +11,37 @@
...
+
+    name: str
+
+    @property
+    def sql(self) -> str:
+        """Returns the unescaped SQL name."""
+        return self.name
+
+    @property
+    def local_normalize(self) -> ColumnId:
```
[the remark this answers is not in the document]
Comment (TrevorBergeron) [reply bigframes/core/identifiers.py]: chnaged to 'normalized'

File: bigframes/core/identifiers.py [hunk 6]
```diff
@@ -11,16 +11,37 @@
...
+        """For use in compiler only. Normalizes to ColumnId referring ot sql name."""
```
[the remark this answers is not in the document]
Comment (TrevorBergeron) [reply bigframes/core/identifiers.py]: fixed

File: bigframes/core/compile/compiled.py (hunk 1 above)
Comment (TrevorBergeron) [reply bigframes/core/compile/compiled.py]: yeah, went a bit overzealous with the nested maps, converted into loop

File: bigframes/core/compile/compiled.py (hunk 2 above)
Comment (TrevorBergeron) [reply bigframes/core/compile/compiled.py]: converted to loop

**Pull request merged**